### PR TITLE
Format: correctly indent block of indented method chain

### LINKED ISSUE #5193: Formatter indentation bug
Extracted from #5146.

```cr
[1, 2, 3]
  .each do |n|
    n
  end
```

Gets formatted to:

```cr
[1, 2, 3]
  .each do |n|
  n
end
```

/cc @MakeNowJust

### PR DESCRIPTION
**Pull request by MakeNowJust**
Fixed #5193

### CODE REVIEW

**Comment by oprypin** [comment]
I'm not sure what you mean by "correctly". I think this is entirely invalid and looks atrocious.
I did not see any calls for action, rather the opposite from the core team.

> In the end, it boils down to user preference, something that the formatter doesn't allow.

Disallowing such chaining would probably be best for readability.

Also here's a more widely accepted perception of "correct":
https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions

**Comment by straight-shoota** [comment]
I think this is much more readable. Just look at the reformatted code in `src/compiler/crystal/tools/expand.cr`. In the current version it is difficult to spot where each indentation level belongs to. This is better with this PR.

Obviously such pieces should be formatted differently to be really good, but this is one step in the right direction.

This would be much cleaner:
```crystal
expansion
  .expanded_sources
  .zip(expansion.expanded_macros)
  .each_with_index do |(expanded_source, expanded_macro), j|		                                    
    expanded_macro.each do |a_macro|
      name = a_macro[:name]
      # ...
    end
    io << "~> "
  end
```

I'm not sure what you mean with *such chaining* exactly, but I'm pretty confident that it shouldn't be generally disallowed.

**Comment by asterite** [comment]
I actually think it's less readable. Before everything fit in 80 columns, now you have to scroll right to see what's going on.

The current formatting is just fine.

I know, it's a personal opinion, and that's why we will never agree on this.

**Comment by RX14** [comment]
I also think this is worse. And I'm also glad crystals code base only changes formatting in one file, it means we don't have much code that's messy in this way.

**Comment by Sija** [comment]
>I know, it's a personal opinion, and that's why we will never agree on this.

@asterite let me ask you a question then - which version would you prefer to write/see?

```cr
expansion.expanded_sources.zip(expansion.expanded_macros)
                          .each_with_index do |(expanded_source, expanded_macro), j|
  expanded_macro.each do |a_macro|
    # ...
  end
end
```

or

```cr
expansion.expanded_sources
  .zip(expansion.expanded_macros)
  .each_with_index do |(expanded_source, expanded_macro), j|
    expanded_macro.each do |a_macro|
      # ...
    end
  end
```

What I'm heading at is placement of `.zip(expansion.expanded_macros)`, which in this case was IMO too far out (pun intended) ;)

File: src/compiler/crystal/tools/expand.cr [hunk 1]
```diff
@@ -21,16 +21,16 @@ module Crystal
         io.puts
         expansion.expanded_sources.zip(expansion.expanded_macros)
                                   .each_with_index do |(expanded_source, expanded_macro), j|
```
Comment (MakeNowJust) [inline src/compiler/crystal/tools/expand.cr]: I should refactor this code in the future, but formatter is not wrong. I believe this indentation is correct and old indent is bad, as @straight-shoota and @Sija said.

**Comment by RX14** [comment]
Thinking about it again, I think @Sija's right.

I think that
```cr
expansion.expanded_sources.zip(expansion.expanded_macros)
                          .each_with_index do |(expanded_source, expanded_macro), j|
  expanded_macro.each do |a_macro|
    # ...
  end
end
```
should format to
```cr
expansion.expanded_sources.zip(expansion.expanded_macros)
    .each_with_index do |(expanded_source, expanded_macro), j|
      expanded_macro.each do |a_macro|
        # ...
      end
    end
```
personally (4 spaces to indent a method chain broken onto multiple lines).

I'd then be encouraged to do this, which I believe to be my favourite formatting:
```cr
expansion.expanded_sources
    .zip(expansion.expanded_macros)
    .each_with_index do |(expanded_source, expanded_macro), j|
      expanded_macro.each do |a_macro|
        # ...
      end
    end
```

**Comment by MakeNowJust** [comment]
I totally agree with @RX14's idea. Dot based alignment is painful. (and why align to last dot on previous line?)

**Comment by straight-shoota** [comment]
@RX14 I'm totally fine with 4 spaces indentation for method chained dot-lines. Though I'm not sure if this needs to be enforced by the formatter.

**Comment by MakeNowJust** [comment]
@asterite Why does the formatter align method chain with the last dot on previous line? So, why the formatter keeps this:

```crystal
foo.bar.baz
       .quz
```

This behavior makes too deep indentation and I think it is wrong. If the formatter aligns method chain with first dot, this PR fix does not lose readability in many case.

**Comment by asterite** [comment]
By now I think the formatter should always use 2 spaces to indent.

**Comment by RX14** [comment]
@asterite I agree normally, but I think that's wrong in the case of multi-line statements. Adding more indent makes it clear that it's a continuation of the previous statement, not just a normal control-flow indent.

**Comment by MakeNowJust** [comment]
I implemented @RX14's idea at [`fix/crystal-format/multiline-call-4-space-indent`](https://github.com/MakeNowJust/crystal/commits/fix/crystal-format/multiline-call-4-space-indent) branch. And [this](https://github.com/MakeNowJust/crystal/commit/d9b2b7f7a15b487fbda681566a1fa5c3d0a1a8c5) is the diff to apply a new `crystal tool format` to crystal source tree. It looks totally good, especially [these lines](https://github.com/MakeNowJust/crystal/commit/d9b2b7f7a15b487fbda681566a1fa5c3d0a1a8c5#diff-a1b7b85341f5f80709d73955200f851eL152) become quite good.

What do you think?

**Comment by MakeNowJust** [comment]
Opened a new pull request #5234.

**Pull request closed without merging**